Developer's Theory Guide for Website

kkiesling opened the pull request:
The PR is for the developer's theory guide for the website. It would be appreciated if people could review the content for correctness.

kkiesling committed:
initialized page for dev theory guide

kkiesling committed:
sense tags added

kkiesling committed:
tweaks

kkiesling committed:
obb images added

kkiesling committed:
ray history added

kkiesling committed:
converted images from svg to png

kkiesling committed:
constructor and additional function calls written up

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+DAGMC works as it exists in MOAB and the key components one may use during
+development.
+
+Contructor
```
Spelling

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+~~~~~~~~~~~~
+
+To initialize a DAGMC instance, the constructor must be called. If the
+constructor is called with no arguments, then a new instance is created. Else,
```
Replace "Else" w/ "Otherwise" perhaps?

pshriwise replied:
There's a need to be specific about what instances are being talked about here. (MOAB instance vs DAGMC instance)

kkiesling replied:
Okay, I think I made the proper clarification, but I am unsure about what exactly is happening. Let me know if how I wrote it up now is correct.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+Topology
+--------
+
+Every geometry contains entities that are either volumes, surfaces, and curves.
```
replace "and" w/ "or" here

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+See sections below for more information about implicit complements, OBBs,
+and indices.
+
+Topology
```
This is probably the most difficult concept to convey and might merit some kind of diagram or tree.
The difference between these geometry entities represented as MOAB entity sets with parent-child relationships to these sets which in turn have a containment relationship with mesh entities that replace the analytic geometric representation. With that in mind I'd be really specific when using the word entity, including some clarifying term in front.

kkiesling replied:
Let's clarify this in person, I am not sure what the best type of diagram is for this clarification.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+Each surface is tagged with the two volume handles of the adjacent volumes.
+The first of the two surfaces is designated as the forward direction and the
+second is designated with the reverse direction. It is important to note that
+these surfaces senses may not be consistent with how an MC code determines
```
Well said in the last sentence.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+Cross Referencing
+-----------------
+
+If one type of indice is known, then either of the other two can be determined.
```
index rather than indice I believe.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+-----------------
+
+If one type of indice is known, then either of the other two can be determined.
+By using the functions below or a combination of the functions, either ID,
```
Global ID for clarity?

kkiesling replied:
yes, thanks for catching that

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+---------------
+
+An OBB, or Oriented Bounding Box, is a box that contains a single facet or set
+of facets and has an axis oriented such that it is the smallest possible box
```
all of the box axes are technically oriented

kkiesling replied:
I think I correctly clarified what you are talking about now, but can you review the new statement and let me know if it is or is not correct now.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+First, the OBB tree for each surface is built. This is built top down where the
+top level is an OBB that contains all the surface’s facets. Then the set of facets
+is split roughly in half and two new OBBs form around each of the new sets.
+This continues until each of the OBBs the level contains a single facet. A 2D
```
I think the default value for number of leaf node facets is 8

pshriwise replied:
Not necessary to alter the existing image though. I'd just be vague and indicate that it contains a small number of facets.

kkiesling replied:
Thanks for clarifying that.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+of the surfaces that particle has crossed is maintained. If a surface encountered
+is reflecting, the particle does not cross that surface and is therefore not
+added to the history. If a particle is or becomes a reflecting particle, the
+history of crossed surfaces is cleared.
```
I think there is more to this construct. @makeclean knows it the best.

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+Otherwise, a random direction is used.
+
+Ray Fire
+~~~~~~~~
```
It is my understanding that this function can only be called for one volume at a time. The bit about ray history here is spot on though.

kkiesling replied:
yes that is correct about only being able to call it one volume at a time and probably a good thing to clarify or note here

pshriwise commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,185 @@
+next volume can be determined by calling next_volume. Given the next surface and
+the known current volume, the next volume is determined by looking at the other
+volume tagged on that surface (as described in the Sense Tags section above).
+This assumes that a valid surface and volume are provided. If no next volumes
```
volumes to volume

pshriwise reviewed (commented):
Mostly minor changes for clarity or typos. This will be very valuable to people trying to understand the code for the first time. Thanks!

ljacobson64 commented:
Just in terms of style, I think it would be helpful if names of functions were formatted as code; i.e. `point_in_volume` instead of point_in_volume

kkiesling committed:
fixed constructor section, maybe

kkiesling committed:
fixed cross-ref section and OBB section

kkiesling committed:
other minor fixes

kkiesling committed:
changed function names to code formatting

kkiesling commented:
Thanks so much @Pshriwise and @ljacobson64 for the initial review! I believe I have addressed all of your comments except for clarifying the topology section. I think I may need a bit of discussion in person to determine the best type of diagram to put there. @Pshriwise if you could just check that my changes to the constructor section and OBB section correctly take into account your comments. Thanks again.

kkiesling committed:
fixed topology section per partick's comments

kkiesling committed:
obb changes again and constructor wording changes

kkiesling commented on doc/contribute/devtheory.rst:
```diff
@@ -0,0 +1,189 @@
+corresponds to the Cubit ID. As a result, the set of global IDs does not
+necessarily have to be contiguous. During setup, the global ID is tagged on the
+entity set. The second method is to identify by index. This is an ordinal
+numerical reference for each volume or surface in the geometry. The third method
```
Index starts from 1

kkiesling committed:
specified that index starts at one

kkiesling committed:
fixed a little wording in ray history

gonuke commented:
I think this only needs an update on RayHistory from @makeclean and then it will be ready..

ljacobson64 commented:
It will also need to have the base branch changed to develop and be rebased against develop as well

makeclean committed:
Added description of the ray history class

makeclean committed:
Fixed spelling mistake

kkiesling committed:
reformatted function names to be consistent with rest of doc

ljacobson64 commented:
Thanks @kkiesling!
